File: OptimoveSDK/app/src/main/java/com/optimove/android/optimovemobilesdk/MainActivity.java
package com.optimove.android.optimovemobilesdk;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import com.optimove.android.Optimove;
import com.optimove.android.main.events.OptimoveEvent;
import com.optimove.android.optimobile.InAppDeepLinkHandlerInterface;
import com.optimove.android.optimobile.InAppInboxItem;
import com.optimove.android.optimobile.OptimoveInApp;
import com.optimove.android.preferencecenter.Channel;
import com.optimove.android.preferencecenter.OptimovePreferenceCenter;
import com.optimove.android.preferencecenter.PreferenceUpdate;
import com.optimove.android.preferencecenter.Preferences;
import com.optimove.android.preferencecenter.Topic;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@SuppressLint("SetTextI18n")
public class MainActivity extends AppCompatActivity {

    static final String TAG = "TestAppMainActvity";
    static final String PC_TAG = "OptimovePC";
    private static final int WRITE_EXTERNAL_PERMISSION_REQUEST_CODE = 169;

    private TextView outputTv;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);
        outputTv = findViewById(R.id.userIdTextView);


        OptimoveInApp.getInstance().setDeepLinkHandler(new InAppDeepLinkHandlerInterface() {
            @Override
            public void handle(android.content.Context context, InAppButtonPress buttonPress) {
                Log.d(TAG, "DeepLink handler invoked");
                startActivity(new Intent(MainActivity.this, MainActivity.class)); // or any Activity you have
            }
        });

        this.hideIrrelevantInputs();

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, WRITE_EXTERNAL_PERMISSION_REQUEST_CODE);
        }

        //deferred deep links
        Optimove.getInstance().seeIntent(getIntent(), savedInstanceState);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Log.d(TAG, "MainActivity.onDestroy called");
    }

    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);

        //deferred deep links
        Optimove.getInstance().seeInputFocus(hasFocus);
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
//    runFromWorker(() -> new DeepLinkHandler(intent).extractLinkData(new LinkDataExtractedListener() {
//      @Override
//      public void onDataExtracted(String screenName, Map<String, String> parameters) {
//        Toast.makeText(MainActivity.this, String.format("New Intent called with screed %s, and data: %s",
//            screenName, Arrays.deepToString(parameters.values().toArray(new String[0]))), Toast.LENGTH_SHORT).show();
//      }
//
//      @Override
//      public void onErrorOccurred(LinkDataError error) {
//        Toast.makeText(MainActivity.this, "New Intent called without deep link", Toast.LENGTH_SHORT).show();
//      }
//    }));

        //deferred deep links
        Optimove.getInstance().seeIntent(intent);
    }

    public void reportEvent(View view) {
        if (view == null)
            return;
        outputTv.setText("Reporting Custom Event for Visitor without optional value");
        runFromWorker(() -> Optimove.getInstance().reportEvent(new SimpleCustomEvent()));
        runFromWorker(() -> Optimove.getInstance().reportEvent("Event_No ParaMs     "));
    }

    public void killActivity(View view) {
        if (view == null)
            return;
        Log.d("TAG", "killActivity called");
        finish(); // forces MainActivity.onDestroy() right now
    }

    public void updateUserId(View view) {
        EditText uidInput = findViewById(R.id.userIdInput);
        EditText emailInput = findViewById(R.id.userEmailInput);
        String userId = uidInput.getText().toString();
        String userEmail = emailInput.getText().toString();

        if (userEmail.isEmpty()) {
            outputTv.setText("Calling setUserId");
            Optimove.getInstance().setUserId(userId);
        } else if (userId.isEmpty()) {
            outputTv.setText("Calling setUserEmail");
            Optimove.getInstance().setUserEmail(userEmail);
        } else {
            outputTv.setText("Calling registerUser");
            Optimove.getInstance().registerUser(userId, userEmail);
        }
    }

    public void readInbox(View view) {
        List<InAppInboxItem> items = OptimoveInApp.getInstance().getInboxItems();
        if (items.size() == 0) {
            Log.d(TAG, "no inbox items!");
            return;
        }
        for (int i = 0; i < items.size(); i++) {
            InAppInboxItem item = items.get(i);
            Log.d(TAG, "title: " + item.getTitle() + ", isRead: " + item.isRead());
        }
    }

    public void markInboxAsRead(View view) {
        Log.d(TAG, "mark  all inbox read");

        OptimoveInApp.getInstance().markAllInboxItemsAsRead();
    }

    public void deleteInbox(View view) {

        List<InAppInboxItem> items = OptimoveInApp.getInstance().getInboxItems();
        if (items.size() == 0) {
            Log.d(TAG, "no inbox items!");
            return;
        }
        for (int i = 0; i < items.size(); i++) {
            OptimoveInApp.getInstance().deleteMessageFromInbox(items.get(i));
        }

    }

    // ******************** PC start *********************

    public void getPreferences(View view) {
        OptimovePreferenceCenter.getInstance().getPreferencesAsync((OptimovePreferenceCenter.ResultType result, Preferences preferences) -> {
            switch (result) {
                case ERROR_USER_NOT_SET:
                case ERROR:
                case ERROR_CREDENTIALS_NOT_SET:
                    Log.d(PC_TAG, result.toString());
                    break;
                case SUCCESS: {

                    Log.d(PC_TAG, "configured: " + preferences.getConfiguredChannels().toString());
                    List<Topic> topics = preferences.getCustomerPreferences();
                    for (int i = 0; i < topics.size(); i++) {
                        Topic topic = topics.get(i);
                        Log.d(PC_TAG, topic.getId() + " " + topic.getName() + " " + topic.getSubscribedChannels().toString());
                    }

                    break;
                }
                default:
                    Log.d(PC_TAG, "unknown res type");
            }
        });

    }

    public void setPreferences(View view) {
        OptimovePreferenceCenter.getInstance().getPreferencesAsync((OptimovePreferenceCenter.ResultType result, Preferences preferences) -> {
            switch (result) {
                case ERROR_USER_NOT_SET:
                case ERROR:
                case ERROR_CREDENTIALS_NOT_SET:
                    Log.d(PC_TAG, result.toString());
                    break;
                case SUCCESS: {
                    Log.d(PC_TAG, "loaded prefs for set: good");


                    List<Channel> configuredChannels = preferences.getConfiguredChannels();
                    List<Topic> topics = preferences.getCustomerPreferences();

                    List<PreferenceUpdate> updates = new ArrayList<>();
                    for (int i = 0; i < topics.size(); i++) {
                        updates.add(new PreferenceUpdate(topics.get(i).getId(), configuredChannels.subList(0, 1)));
                    }

                    OptimovePreferenceCenter.getInstance().setCustomerPreferencesAsync((OptimovePreferenceCenter.ResultType setResult) -> {
                        Log.d(PC_TAG, setResult.toString());
                    }, updates);

                    break;
                }
                default:
                    Log.d(PC_TAG, "unknown res type");
            }
        });
    }

    // ******************** PC end *********************

    public void viewEmbeddedMessaging(View view) {
        // Do nothing for now
        Intent intent = new Intent(this, EmbeddedMessagingActivity.class);
        startActivity(intent);
    }

    public void setCredentials(View view) {
        EditText optimoveCreds = findViewById(R.id.optimoveCredInput);
        EditText optimobileCreds = findViewById(R.id.optimobileCredInput);
        EditText pcCreds = findViewById(R.id.prefCenterCredInput);

        String optimoveCredentials = optimoveCreds.getText().toString();
        String optimobileCredentials = optimobileCreds.getText().toString();
        String prefCenterCredentials = pcCreds.getText().toString();

        if (optimoveCredentials.isEmpty() && optimobileCredentials.isEmpty()) {
            return;
        }

        if (optimoveCredentials.isEmpty()) {
            optimoveCredentials = null;
        }

        if (optimobileCredentials.isEmpty()) {
            optimobileCredentials = null;
        }

        if (prefCenterCredentials.isEmpty()) {
            prefCenterCredentials = null;
        }

        try {
            Optimove.setCredentials(optimoveCredentials, optimobileCredentials, prefCenterCredentials);
        } catch (Exception e) {
            outputTv.setText(e.getMessage());
            return;
        }

        outputTv.setText("Credentials submitted");
        Button setCredsBtn = (Button) findViewById(R.id.submitCredentialsBtn);
        setCredsBtn.setEnabled(false);
    }

    public void runFromWorker(Runnable runnable) {
        new Thread(runnable).start();
    }

    private static class SimpleCustomEvent extends OptimoveEvent {

        SimpleCustomEvent() {
        }

        @Override
        public String getName() {
            return "Simple cUSTOM_Event     ";
        }

        @Override
        public Map<String, Object> getParameters() {
            HashMap<String, Object> result = new HashMap<>();
            String val = "  some_string  ";
            result.put("strinG_param", val);
            result.put("number_param", 42);
            return result;
        }
    }

    private void hideIrrelevantInputs() {
        if (!Optimove.getConfig().isPreferenceCenterConfigured()) {
            Button getPrefsBtn = (Button) findViewById(R.id.getPreferences);
            getPrefsBtn.setVisibility(View.GONE);

            Button setPrefsBtn = (Button) findViewById(R.id.setPreferences);
            setPrefsBtn.setVisibility(View.GONE);
        }

        if (!Optimove.getConfig().isEmbeddedMessagingConfigured()) {
            Button viewEmbeddedMessagingBtn = findViewById(R.id.viewEmbeddedMessaging);
            viewEmbeddedMessagingBtn.setVisibility(View.GONE);
        }

        if (!Optimove.getConfig().usesDelayedConfiguration()) {
            EditText optimoveCredInput = findViewById(R.id.optimoveCredInput);
            optimoveCredInput.setVisibility(View.GONE);

            EditText optimobileCredInput = findViewById(R.id.optimobileCredInput);
            optimobileCredInput.setVisibility(View.GONE);

            EditText pcCredInput = findViewById(R.id.prefCenterCredInput);
            pcCredInput.setVisibility(View.GONE);

            Button setCredsBtn = (Button) findViewById(R.id.submitCredentialsBtn);
            setCredsBtn.setVisibility(View.GONE);


        }
    }
}

File: OptimoveSDK/app/src/main/java/com/optimove/android/optimovemobilesdk/MainActivity.kt
package com.optimove.android.optimovemobilesdk

import android.Manifest
import android.annotation.SuppressLint
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.activity.compose.setContent
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import android.content.Context
import android.location.Location
import android.content.SharedPreferences
import android.os.Build
import android.provider.Settings
import java.io.File
import android.os.Handler
import android.os.Looper
import com.optimove.android.Optimove
import com.optimove.android.main.events.OptimoveEvent
import com.optimove.android.optimobile.InAppDeepLinkHandlerInterface
import com.optimove.android.optimobile.InAppMessageInterceptor
import com.optimove.android.optimobile.InAppMessageInterceptorCallback
import com.optimove.android.optimobile.OptimoveInApp
import com.optimove.android.optimovemobilesdk.ui.MainScreen

import com.optimove.android.optimovemobilesdk.ui.theme.AppTheme
import org.json.JSONObject

@SuppressLint("SetTextI18n")
class MainActivity : AppCompatActivity() {

    private var outputText by mutableStateOf("Welcome to the Optimove Android SDK test")
    private var credentialsSubmitted by mutableStateOf(false)
    private var isInterceptingInApp by mutableStateOf(false)
    private var isDelayedInit by mutableStateOf(false)
    private var isAuthEnabled by mutableStateOf(false)
    private var inAppDecisionDialog: AlertDialog? = null
    private var persistedUserId by mutableStateOf("")
    private var persistedUserEmail by mutableStateOf("")
    private lateinit var identityPrefs: SharedPreferences
    private lateinit var qaPrefs: SharedPreferences

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        OptimoveInApp.getInstance().setDeepLinkHandler(object : InAppDeepLinkHandlerInterface {
            override fun handle(
                context: android.content.Context,
                buttonPress: InAppDeepLinkHandlerInterface.InAppButtonPress
            ) {
                Log.d(TAG, "DeepLink handler invoked")
                startActivity(Intent(this@MainActivity, MainActivity::class.java))
            }
        })

        val config = Optimove.getConfig()
        val showPreferenceCenter = config.isPreferenceCenterConfigured
        val showEmbeddedMessaging = config.isEmbeddedMessagingConfigured
        val showOverlayMessaging = config.isOverlayMessagingEnabled
        val showDelayedConfig = config.usesDelayedConfiguration()

        if (ContextCompat.checkSelfPermission(
                this, Manifest.permission.WRITE_EXTERNAL_STORAGE
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(
                this,
                arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
                WRITE_EXTERNAL_PERMISSION_REQUEST_CODE
            )
        }

        Optimove.getInstance().seeIntent(intent, savedInstanceState)
        Optimove.getInstance().pushRequestDeviceToken()

        identityPrefs = getSharedPreferences(IDENTITY_PREF_NAME, Context.MODE_PRIVATE)
        persistedUserEmail = identityPrefs.getString(KEY_USER_EMAIL, "") ?: ""
        persistedUserId = identityPrefs.getString(KEY_USER_ID, "") ?: ""

        qaPrefs = getSharedPreferences(MyApplication.PREFS_NAME, Context.MODE_PRIVATE)
        isDelayedInit = qaPrefs.getBoolean(MyApplication.KEY_DELAYED_INIT, true)
        isAuthEnabled = qaPrefs.getBoolean(MyApplication.KEY_AUTH_ENABLED, false)

        setContent {
            AppTheme {
                MainScreen(
                    outputText = outputText,
                    showPreferenceCenter = showPreferenceCenter,
                    showEmbeddedMessaging = showEmbeddedMessaging,
                    showOverlayMessaging = showOverlayMessaging,
                    showDelayedConfig = showDelayedConfig,
                    credentialsSubmitted = credentialsSubmitted,
                    isInterceptingInApp = isInterceptingInApp,
                    userId = persistedUserId,
                    userEmail = persistedUserEmail,
                    onUserIdChange = { persistedUserId = it },
                    onUserEmailChange = { persistedUserEmail = it },
                    onReportEvent = ::reportEvent,
                    onClearAppData = ::clearAppData,
                    onKillActivity = ::killActivity,
                    onUpdateUserId = ::updateUserId,
                    onClearIdentity = ::clearIdentity,
                    onReadInbox = ::readInbox,
                    onMarkInboxAsRead = ::markInboxAsRead,
                    onDeleteInbox = ::deleteInbox,
                    onViewPreferenceCenter = ::viewPreferenceCenter,
                    onViewEmbeddedMessaging = ::viewEmbeddedMessaging,
                    onViewOverlayMessaging = ::viewOverlayMessaging,
                    onSetCredentials = ::setCredentials,
                    onInAppInterceptionClicked = ::onInAppInterceptionClicked,
                    onSendLocation = { lat, lng ->
                        val location = Location("test").apply {
                            latitude = lat
                            longitude = lng
                            time = System.currentTimeMillis()
                        }
                        Optimove.getInstance().sendLocationUpdate(location)
                        outputText = "Location sent: ($lat, $lng)"
                    },
                    onResetToken = {},
                    onOpenDeeplinkTest = ::openDeeplinkTest,
                    onRegisterPush = {
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU && ContextCompat.checkSelfPermission(
                                this@MainActivity, Manifest.permission.POST_NOTIFICATIONS
                            ) == PackageManager.PERMISSION_DENIED && !shouldShowRequestPermissionRationale(
                                Manifest.permission.POST_NOTIFICATIONS
                            )
                        ) {
                            outputText =
                                "Notification permission permanently denied. Opening settings..."
                            startActivity(Intent(Settings.ACTION_APP_NOTIFICATION_SETTINGS).apply {
                                putExtra(Settings.EXTRA_APP_PACKAGE, this@MainActivity.packageName)
                            })
                        } else {
                            outputText = "Requesting push registration..."
                            Optimove.getInstance().pushRequestDeviceToken()
                            outputText = "Push registration requested"
                        }
                    },
                    onUnregisterPush = {
                        outputText = "Unregistering push..."
                        Optimove.getInstance().pushUnregister()
                        outputText = "Push unregistration requested"
                    },
                    isDelayedInit = isDelayedInit,
                    onDelayedInitToggle = { enabled ->
                        isDelayedInit = enabled
                        qaPrefs.edit().putBoolean(MyApplication.KEY_DELAYED_INIT, enabled).apply()

                        outputText = if (enabled)
                            "Delayed init enabled — restart app to apply"
                        else
                            "Immediate init enabled — restart app to apply"
                    },
                    isAuthEnabled = isAuthEnabled,
                    onAuthToggle = { enabled ->
                        isAuthEnabled = enabled
                        qaPrefs.edit().putBoolean(MyApplication.KEY_AUTH_ENABLED, enabled).apply()

                        outputText = if (enabled)
                            "Auth enabled — restart app to apply"
                        else
                            "Auth disabled — restart app to apply"
                    },
                    onOpenAuthDebug = ::openAuthDebug,
                    onOpenGamifyWidget = ::openGamifyWidget
                )
            }
        }
    }

    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        Optimove.getInstance().seeInputFocus(hasFocus)
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        Optimove.getInstance().seeIntent(intent)
    }

    override fun onDestroy() {
        inAppDecisionDialog?.takeIf { it.isShowing }?.dismiss()
        super.onDestroy()
    }

    private fun reportEvent() {
        outputText = "Reporting Custom Event for Visitor without optional value"
        runFromWorker { Optimove.getInstance().reportEvent(SimpleCustomEvent()) }
        runFromWorker { Optimove.getInstance().reportEvent("Event_No ParaMs     ") }
    }

    private fun clearAppData() {
        AlertDialog.Builder(this)
            .setTitle("Clear App Data")
            .setMessage(
                "This deletes all local app data and restarts the app."
            )
            .setPositiveButton("Clear") { _, _ ->
                Thread {
                    wipePrivateAppDataFilesystem()
                    runOnUiThread {
                        val launch = packageManager.getLaunchIntentForPackage(packageName)
                        if (launch != null) {
                            launch.addFlags(
                                Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
                            )
                            startActivity(launch)
                        }
                        Handler(Looper.getMainLooper()).postDelayed({
                            Runtime.getRuntime().exit(0)
                        }, 150)
                    }
                }.start()
            }
            .setNegativeButton("Cancel", null)
            .show()
    }

    private fun wipePrivateAppDataFilesystem() {
        val ctx = applicationContext
        val dataDir = File(ctx.applicationInfo.dataDir)
        dataDir.listFiles()?.forEach { child ->
            runCatching { child.deleteRecursively() }
                .onFailure { Log.w(TAG, "Failed to delete ${child.path}", it) }
        }
        runCatching {
            ctx.getExternalFilesDir(null)?.parentFile?.deleteRecursively()
        }.onFailure { Log.w(TAG, "Failed to delete external app data dir", it) }
    }

    private fun killActivity() {
        Log.d("TAG", "killActivity called")
        finish()
    }

    private fun updateUserId(userId: String, userEmail: String) {
        when {
            userEmail.isEmpty() -> {
                outputText = "Calling setUserId"
                Optimove.getInstance().setUserId(userId)
            }

            userId.isEmpty() -> {
                outputText = "Calling setUserEmail"
                Optimove.getInstance().setUserEmail(userEmail)
            }

            else -> {
                outputText = "Calling registerUser"
                Optimove.getInstance().registerUser(userId, userEmail)
            }
        }
        identityPrefs.edit().putString(KEY_USER_ID, userId).putString(KEY_USER_EMAIL, userEmail)
            .apply()
    }

    private fun clearIdentity() {
        identityPrefs.edit().remove(KEY_USER_ID).remove(KEY_USER_EMAIL).apply()
        persistedUserId = ""
        persistedUserEmail = ""
        outputText = "Identity cleared (saved values removed)"
    }

    private fun readInbox() {
        val items = OptimoveInApp.getInstance().inboxItems
        if (items.isEmpty()) {
            Log.d(TAG, "no inbox items!")
            return
        }
        items.forEach { item ->
            Log.d(TAG, "title: ${item.title}, isRead: ${item.isRead}")
        }
    }

    private fun markInboxAsRead() {
        Log.d(TAG, "mark  all inbox read")
        OptimoveInApp.getInstance().markAllInboxItemsAsRead()
    }

    private fun deleteInbox() {
        val items = OptimoveInApp.getInstance().inboxItems
        if (items.isEmpty()) {
            Log.d(TAG, "no inbox items!")
            return
        }
        items.forEach { OptimoveInApp.getInstance().deleteMessageFromInbox(it) }
    }


    private fun viewPreferenceCenter() {
        startActivity(Intent(this, PreferenceCenterActivity::class.java))
    }

    private fun openAuthDebug() {
        startActivity(Intent(this, AuthDebugActivity::class.java))
    }

    private fun viewEmbeddedMessaging() {
        startActivity(Intent(this, EmbeddedMessagingActivity::class.java))
    }


    private fun viewOverlayMessaging() {
        startActivity(Intent(this, OverlayMessagingActivity::class.java))
    }

    private fun openGamifyWidget() {
        startActivity(Intent(this, GamifyWidgetActivity::class.java))

    }

    private fun openDeeplinkTest() {
        startActivity(
            Intent(
                Intent.ACTION_VIEW, Uri.parse(DeeplinkTargetActivity.DEEPLINK_TEST_URI)
            )
        )
    }

    private fun setCredentials(optimove: String?, optimobile: String?, prefCenter: String?) {
        if (optimove.isNullOrEmpty() && optimobile.isNullOrEmpty()) return
        try {
            Optimove.setCredentials(optimove, optimobile, prefCenter)
        } catch (e: Exception) {
            outputText = e.message ?: "Error"
            return
        }
        outputText = "Credentials submitted"
        credentialsSubmitted = true
    }

    private fun runFromWorker(runnable: () -> Unit) {
        Thread(runnable).start()
    }

    private fun onInAppInterceptionClicked() {
        if (isInterceptingInApp) {
            disableInAppInterception()
        } else {
            showEnableInAppInterceptionDialog()
        }
    }

    private fun showEnableInAppInterceptionDialog() {
        val options = arrayOf("Default (5000 ms)", "12,000 ms")
        var selected = 0

        AlertDialog.Builder(this).setTitle("Enable In-App Interception")
            .setSingleChoiceItems(options, 0) { _, which -> selected = which }
            .setPositiveButton("Enable") { d, _ ->
                val timeoutMs = if (selected == 0) 5000L else 12000L
                enableInAppInterception(timeoutMs)
                isInterceptingInApp = true

                outputText = "In-App interception enabled ($timeoutMs ms timeout per message)"
                Toast.makeText(
                    this,
                    "In-App interception enabled ($timeoutMs ms)",
                    Toast.LENGTH_SHORT
                ).show()

                d.dismiss()
            }.setNegativeButton("Cancel", null).show()
    }

    private fun disableInAppInterception() {
        inAppDecisionDialog?.takeIf { it.isShowing }?.dismiss()
        OptimoveInApp.getInstance().setInAppMessageInterceptor(null)
        isInterceptingInApp = false
        outputText = "In-App interception disabled"
        Toast.makeText(this, "In-App interception disabled", Toast.LENGTH_SHORT).show()
    }

    private fun enableInAppInterception(timeoutMs: Long) {
        OptimoveInApp.getInstance().setInAppMessageInterceptor(object : InAppMessageInterceptor {
            override fun processMessage(
                messageData: JSONObject?, decision: InAppMessageInterceptorCallback
            ) {
                runOnUiThread {
                    inAppDecisionDialog?.takeIf { it.isShowing }?.dismiss()
                    val dataText = messageData?.toString() ?: "No data provided"

                    inAppDecisionDialog = AlertDialog.Builder(this@MainActivity)
                        .setTitle("QA: In-App Message")
                        .setMessage("$dataText\nChoose Show, Postpone, or Suppress.")
                        .setPositiveButton("Show") { dialog, _ ->
                            decision.show()
                            dialog.dismiss()
                        }
                        .setNeutralButton("Postpone") { dialog, _ ->
                            decision.postpone()
                            dialog.dismiss()
                        }
                        .setNegativeButton("Suppress") { dialog, _ ->
                            decision.suppress()
                            dialog.dismiss()
                        }
                        .setOnCancelListener { decision.suppress() }
                        .create()

                    inAppDecisionDialog?.show()
                }
            }

            override fun getTimeoutMs(): Long = timeoutMs
        })
    }

    private class SimpleCustomEvent : OptimoveEvent() {
        override fun getName(): String = "Simple cUSTOM_Event     "
        override fun getParameters(): Map<String, Any> = mapOf(
            "strinG_param" to "  some_string  ", "number_param" to 42
        )
    }

    companion object {
        private const val TAG = "TestAppMainActvity"
        private const val OVERLAY_TAG = "OverlayMessaging"
        private const val WRITE_EXTERNAL_PERMISSION_REQUEST_CODE = 169
        private const val IDENTITY_PREF_NAME = "optimove_identity"
        private const val KEY_USER_EMAIL = "user_email"
        private const val KEY_USER_ID = "user_id"
    }
}
